File: Samples/GridLoaders/DNN/VB/GridLoader/GridBase.vb
Imports System
Imports System.Data
Imports System.Configuration
Imports System.Web
Imports System.Web.Security
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Web.UI.WebControls.WebParts
Imports System.Web.UI.HtmlControls

Imports EntitySpaces.DynamicQuery
Imports EntitySpaces.Interfaces

Imports DotNetNuke.Common
Imports DotNetNuke.Entities.Modules

Namespace BusinessObjects

    Public MustInherit Class GridBase
        Inherits PortalModuleBase

        Protected Const BrowseView As Integer = 0
        Protected Const DetailView As Integer = 1
        Protected Const EditView As Integer = 2
        Protected Const SearchView As Integer = 3

        Protected MustOverride Sub DisplayAddView()
        Protected MustOverride Sub DisplayDetailView()
        Protected MustOverride Sub DisplayEditView()
        Protected MustOverride Sub DisplaySearchView()

        Protected MustOverride Sub Save()
        Protected MustOverride Sub LoadGrid()
        Protected MustOverride Sub Delete()
        Protected MustOverride Sub ResetEditFields()
        Protected MustOverride Sub ClearSelectedKeys()
        Protected MustOverride Function HasSelectedKeys() As Boolean

        Protected MustOverride ReadOnly Property BrowseGrid() As GridView
        Protected MustOverride ReadOnly Property MultiView() As MultiView
        Protected MustOverride ReadOnly Property ShowAllButton() As Button
        Protected MustOverride ReadOnly Property ClearSearchCriteriaButton() As Button
        Protected MustOverride ReadOnly Property DeleteErrorLabel() As Label
        Protected MustOverride ReadOnly Property EditHeaderLabel() As Label
        Protected MustOverride ReadOnly Property SaveErrorLabel() As Label

        Protected MustOverride ReadOnly Property PageTitle() As String

#Region "Viewstate Properties"
        Protected Property IsNew() As Boolean
            Get
                Dim o As Object = Me.ViewState("IsNew")

                If Not o Is Nothing Then
                    Return CType(o, Boolean)
                Else
                    Return False
                End If
            End Get
            Set(ByVal value As Boolean)
                Me.ViewState("IsNew") = value
            End Set
        End Property

        Protected Property SearchHasBeenLoaded() As Boolean
            Get
                Dim o As Object = Me.ViewState("Search")

                If Not o Is Nothing Then
                    Return CType(o, Boolean)
                Else
                    Return False
                End If
            End Get
            Set(ByVal value As Boolean)
                Me.ViewState("Search") = value
            End Set
        End Property

#End Region

#Region "Button Handlers"

        Protected Sub bnNew_Click(ByVal sender As Object, ByVal e As System.EventArgs)
            Me.MultiView.ActiveViewIndex = EditView
            Me.DisplayAddView()
        End Sub

        Protected Sub bnShowAll_Click(ByVal sender As Object, ByVal e As System.EventArgs)
            If (Me.BrowseGrid.AllowPaging) Then
                Me.BrowseGrid.AllowPaging = False
                Me.ShowAllButton.Text = Me.BrowseGrid.PageSize.ToString() + " Rows"
            Else
                Me.BrowseGrid.AllowPaging = True
                Me.ShowAllButton.Text = "Show All"
            End If

            Me.LoadGrid()
        End Sub

        Protected Sub bnBrowse_Click(ByVal sender As Object, ByVal e As System.EventArgs)
            Response.Redirect(String.concat(Globals.ApplicationPath, "/tabid/", TabId, "/pagename/", PageTitle, "/default.aspx"))
        End Sub

        Protected Sub bnEdit_Click(ByVal sender As Object, ByVal e As System.EventArgs)
            Me.MultiView.ActiveViewIndex = EditView
            DisplayEditView()
        End Sub

        Protected Sub bnDelete_Click(ByVal sender As Object, ByVal e As System.EventArgs)

            Try
                Delete()
                Me.MultiView.ActiveViewIndex = BrowseView
                LoadGrid()
            Catch ex As Exception
                Me.DeleteErrorLabel.Text = "Unable to Delete this Record"
            End Try

        End Sub

        Protected Sub bnSaveNew_Click(ByVal sender As Object, ByVal e As System.EventArgs)

            Try
                Me.MultiView.ActiveViewIndex = EditView
                Save()
                Me.EditHeaderLabel.Text = "New " + Me.PageTitle
                Me.ClearSelectedKeys()
                Me.IsNew = True
                Me.ResetEditFields()
            Catch ex As Exception
                Me.SaveErrorLabel.Text = ex.Message
            End Try

        End Sub

        Protected Sub bnCancel_Click(ByVal sender As Object, ByVal e As System.EventArgs)
            If Not Request.QueryString("return") = Nothing Then
                Response.Redirect(Request.QueryString("return"))
            Else
                If Me.HasSelectedKeys() Then
                    Me.MultiView.ActiveViewIndex = DetailView
                Else
                    Me.MultiView.ActiveViewIndex = BrowseView
                End If
            End If
        End Sub

        Protected Sub SelectedIndexChanged(ByVal grid As GridView, ByVal pageName As String)

            Dim selectedRow As Integer = grid.SelectedIndex
            Dim key As DataKey = grid.DataKeys(selectedRow)

            Dim seperator As String = ""
            Dim urlParams As String = ""
            Dim keyName As String

            For Each keyName In grid.DataKeyNames
                urlParams += seperator
                urlParams += keyName + "/"
                urlParams += Convert.ToString(key(keyName))
                seperator = "/"
            Next

            Response.Redirect(String.Concat(Globals.ApplicationPath, "/tabid/", TabId, "/pagename/", pageName, "/", urlParams, "/default.aspx"))
        End Sub

        Protected Sub gvBrowse_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs)
            Me.SelectedIndexChanged(Me.BrowseGrid, Me.PageTitle)
        End Sub

        Protected Sub gvBrowse_PageIndexChanging(ByVal sender As Object, ByVal e As GridViewPageEventArgs)
            Me.BrowseGrid.PageIndex = e.NewPageIndex
            LoadGrid()
        End Sub

        Protected Sub gvBrowse_Sorting(ByVal sender As Object, ByVal e As GridViewSortEventArgs)
            Me.SetSortInformation("gvBrowse", e.SortExpression)
            Me.LoadGrid()
        End Sub

        Protected Sub bnSearchPage_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.MultiView.ActiveViewIndex = SearchView
            DisplaySearchView()
            Me.SearchHasBeenLoaded = True
        End Sub

        Protected Sub bnSearch_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.MultiView.ActiveViewIndex = BrowseView
            Me.LoadGrid()
            Me.ClearSearchCriteriaButton.Visible = True
        End Sub

#End Region

#Region "Sorting Logic"

        Protected Function SetSortInformation(ByVal gridName As String, ByVal expression As String) As Boolean

            Dim changed As Boolean = False

            Dim o As Object = Me.ViewState(gridName + "EXP")
            If Not o Is Nothing Then
                If Not CType(o, String) = expression Then
                    changed = True
                End If
            Else
                changed = True
            End If

            Me.ViewState(gridName + "EXP") = expression

            If changed = True Then
                Me.ViewState(gridName + "DIR") = "Ascending"
            Else
                Select Case CType(Me.ViewState(gridName + "DIR"), String)
                    Case "Ascending"
                        Me.ViewState(gridName + "DIR") = "Descending"
                    Case "Descending"
                        Me.ViewState(gridName + "DIR") = "Ascending"
                    Case Else
                        Me.ViewState(gridName + "DIR") = "Ascending"
                End Select
            End If

            Return changed

        End Function

        Protected Function GetSortExpression(ByVal gridName As String) As String
            Dim o As Object = Me.ViewState(gridName + "EXP")
            Return CStr(IIf(Not o Is Nothing, CType(o, String), ""))
        End Function

        Protected Sub SetSortExpression(ByVal gridName As String, ByVal expression As String)
            Me.ViewState(gridName + "EXP") = expression
        End Sub

        Protected Function GetSortDirection(ByVal gridName As String) As String
            Dim o As Object = Me.ViewState(gridName + "DIR")
            Return CStr(IIf(Not o Is Nothing, CType(o, String), ""))
        End Function

        Protected Function GetSortDirectionEnum(ByVal gridName As String) As esOrderByDirection

            Dim dir As esOrderByDirection = esOrderByDirection.Ascending

            Dim direction As String = CType(Me.ViewState(gridName + "DIR"), String)
            If Not direction = "" Then
                Select Case direction
                    Case "Ascending"
                        dir = esOrderByDirection.Ascending
                    Case "Descending"
                        dir = esOrderByDirection.Descending
                End Select
            End If

            Return dir

        End Function

        Protected Sub SetSortDirection(ByVal gridName As String, ByVal expression As String)
            Me.ViewState(gridName + "DIR") = expression
        End Sub

#End Region

        Protected Function PopulateSearchItemList() As ListItem()

            Dim items As ListItem() = { _
                New ListItem("", ""), _
                New ListItem("Equal", "Equal"), _
                New ListItem("Not Equal", "NotEqual"), _
                New ListItem("Greater Than", "GreaterThan"), _
                New ListItem("Greater Than Or Equal", "GreaterThanOrEqual"), _
                New ListItem("Less Than", "LessThan"), _
                New ListItem("Less Than Or Equal", "LessThanOrEqual"), _
                New ListItem("Like", "Like"), _
                New ListItem("Is Null", "IsNull"), _
                New ListItem("Is Not Null", "IsNotNull"), _
                New ListItem("Not Like", "NotLike")}

            Return items

        End Function

    End Class

End Namespace

File: Samples/GridLoaders/DNN/C#/GridLoader/GridBase.cs
using System;
using System.Data;
using System.Configuration;
using System.Web;
using System.Web.Security;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;

using EntitySpaces.DynamicQuery;
using EntitySpaces.Interfaces;

using DotNetNuke.Common;
using DotNetNuke.Entities.Modules;

namespace BusinessObjects
{
	abstract public class GridBase : PortalModuleBase
	{
		public GridBase() {}

		protected const int BrowseView = 0;
		protected const int DetailView = 1;
		protected const int EditView   = 2;
		protected const int SearchView = 3;

		abstract protected void DisplayAddView();
		abstract protected void DisplayDetailView();
		abstract protected void DisplayEditView();
		abstract protected void DisplaySearchView();

		abstract protected void Save();
		abstract protected void LoadGrid();
		abstract protected void Delete();
		abstract protected void ResetEditFields();
		abstract protected void ClearSelectedKeys();
		abstract protected bool HasSelectedKeys();

		abstract protected GridView BrowseGrid { get; }
		abstract protected MultiView MultiView { get; }
		abstract protected Button ShowAllButton { get; }
		abstract protected Button ClearSearchCriteriaButton { get; }
		abstract protected Label DeleteErrorLabel { get; }
		abstract protected Label EditHeaderLabel { get; }
		abstract protected Label SaveErrorLabel { get; }

		abstract protected string PageTitle { get; }

		#region Viewstate Properties
		protected bool IsNew
		{
			get
			{
				object o = this.ViewState["IsNew"];

				if (o != null)
					return (bool)o;
				else
					return false;
			}

			set
			{
				this.ViewState["IsNew"] = value;
			}
		}

		protected bool SearchHasBeenLoaded
		{
			get
			{
				object o = this.ViewState["Search"];

				if (o != null)
					return (bool)o;
				else
					return false;
			}

			set
			{
				this.ViewState["Search"] = value;
			}
		}

		#endregion

		#region Button Handlers

		protected void bnNew_Click(object sender, EventArgs e)
		{
			this.MultiView.ActiveViewIndex = EditView;
			this.DisplayAddView();
		}

		protected void bnShowAll_Click(object sender, EventArgs e)
		{
			if (this.BrowseGrid.AllowPaging)
			{
				this.BrowseGrid.AllowPaging = false;
				this.ShowAllButton.Text = this.BrowseGrid.PageSize.ToString() + " Rows";
			}
			else
			{
				this.BrowseGrid.AllowPaging = true;
				this.ShowAllButton.Text = "Show All";
			}
			this.LoadGrid();
		}

		protected void bnBrowse_Click(object sender, EventArgs e)
		{
			Response.Redirect(string.Concat(Globals.ApplicationPath, "/tabid/", TabId, 
				"/pagename/", PageTitle, "/default.aspx")); 
		}

		protected void bnEdit_Click(object sender, EventArgs e)
		{
			this.MultiView.ActiveViewIndex = EditView;
			DisplayEditView();
		}

		protected void bnDelete_Click(object sender, EventArgs e)
		{
			try
			{
				Delete();
				this.MultiView.ActiveViewIndex = BrowseView;
				LoadGrid();
			}
			catch (Exception ex)
			{
				this.DeleteErrorLabel.Text = "Unable to Delete this Record";
			}
		}

		protected void bnSaveNew_Click(object sender, EventArgs e)
		{
			try
			{
				this.MultiView.ActiveViewIndex = EditView;
				Save();
				this.EditHeaderLabel.Text = "New " + this.PageTitle;
				this.ClearSelectedKeys();
				this.IsNew = true;
				this.ResetEditFields();
			}
			catch (Exception ex)
			{
				this.SaveErrorLabel.Text = ex.Message;
			}
		}

		protected void bnCancel_Click(object sender, EventArgs e)
		{
			if (Request.QueryString["return"] != null)
			{
				Response.Redirect(Request.QueryString["return"]);
			}
			else
			{
				if (this.HasSelectedKeys())
					this.MultiView.ActiveViewIndex = DetailView;
				else
					this.MultiView.ActiveViewIndex = BrowseView;
			}
		}

		protected void SelectedIndexChanged(GridView grid, string pageName)
		{
			int selectedRow = grid.SelectedIndex;
			DataKey key = grid.DataKeys[selectedRow];

			string seperator = "";
			string urlParams = "";

			foreach (string keyName in grid.DataKeyNames)
			{
				urlParams += seperator;
				urlParams += keyName + "/";
				urlParams += Convert.ToString(key[keyName]);
				seperator = "/";
			}

			Response.Redirect(string.Concat(Globals.ApplicationPath, "/tabid/", TabId,
				"/pagename/", pageName, "/", urlParams, "/default.aspx"));
		}

		protected void gvBrowse_SelectedIndexChanged(object sender, EventArgs e)
		{
			this.SelectedIndexChanged(this.BrowseGrid, this.PageTitle);
		}

		protected void gvBrowse_PageIndexChanging(object sender, GridViewPageEventArgs e)
		{
			this.BrowseGrid.PageIndex = e.NewPageIndex;
			LoadGrid();
		}

		protected void gvBrowse_Sorting(object sender, GridViewSortEventArgs e)
		{
			this.SetSortInformation("gvBrowse", e.SortExpression);
			this.LoadGrid();
		}

		protected void bnSearchPage_Click(object sender, EventArgs e)
		{
			this.MultiView.ActiveViewIndex = SearchView;
			DisplaySearchView();
			this.SearchHasBeenLoaded = true;
		}

		protected void bnSearch_Click(object sender, EventArgs e)
		{
			this.MultiView.ActiveViewIndex = BrowseView;
			this.LoadGrid();
			this.ClearSearchCriteriaButton.Visible = true;
		}

		#endregion

		#region Sorting Logic
		protected bool SetSortInformation(string gridName, string expression)
		{
			bool changed = false;

			object o = this.ViewState[gridName + "EXP"];
			if (o != null)
			{
				if (o as string != expression)
				{
					changed = true;
				}
			}
			else
			{
				changed = true;
			}

			this.ViewState[gridName + "EXP"] = expression;

			if (changed)
			{
				this.ViewState[gridName + "DIR"] = "Ascending";
			}
			else
			{
				switch (this.ViewState[gridName + "DIR"] as string)
				{
					case "Ascending":
						this.ViewState[gridName + "DIR"] = "Descending";
						break;

					case "Descending":
						this.ViewState[gridName + "DIR"] = "Ascending";
						break;

					default:
						this.ViewState[gridName + "DIR"] = "Ascending";
						break;
				}
			}

			return changed;
		}

		protected string GetSortExpression(string gridName)
		{
			object o = this.ViewState[gridName + "EXP"];
			return (o != null) ? o as string : "";
		}

		protected void SetSortExpression(string gridName, string expression)
		{
			this.ViewState[gridName + "EXP"] = expression;
		}

		protected string GetSortDirection(string gridName)
		{
			object o = this.ViewState[gridName + "DIR"];
			return (o != null) ? o as string : "";
		}

		protected esOrderByDirection GetSortDirectionEnum(string gridName)
		{
			string direction = this.ViewState[gridName + "DIR"] as string;
			if (direction != "")
			{
				switch (direction)
				{
					case "Ascending": return esOrderByDirection.Ascending;
					case "Descending": return esOrderByDirection.Descending;
				}
			}

			return esOrderByDirection.Ascending;
		}

		protected void SetSortDirection(string gridName, string expression)
		{
			this.ViewState[gridName + "DIR"] = expression;
		}
		#endregion

		protected ListItem[] PopulateSearchItemList()
		{
			ListItem[] items = new ListItem[11];
			items[0] = new ListItem("", "");
			items[1] = new ListItem("Equal", "Equal");
			items[2] = new ListItem("Not Equal", "NotEqual");
			items[3] = new ListItem("Greater Than", "GreaterThan");
			items[4] = new ListItem("Greater Than Or Equal", "GreaterThanOrEqual");
			items[5] = new ListItem("Less Than", "LessThan");
			items[6] = new ListItem("Less Than Or Equal", "LessThanOrEqual");
			items[7] = new ListItem("Like", "Like");
			items[8] = new ListItem("Is Null", "IsNull");
			items[9] = new ListItem("Is Not Null", "IsNotNull");
			items[10] = new ListItem("Not Like", "NotLike");
			return items;
		}
	}
}